File: swift/Sources/DeviceAiLlm/Session/ChatSession.swift
import Foundation
import DeviceAiCore

/// A stateful LLM conversation session.
///
/// Create via `DeviceAI.llm.chat(modelPath:)` — never instantiate directly.
///
/// ## Usage
///
/// ```swift
/// let session = DeviceAI.llm.chat(modelPath: "/path/to/model.gguf") {
///     $0.systemPrompt = "You are a helpful assistant."
///     $0.temperature  = 0.8
/// }
///
/// // Streaming (recommended for UI — users see tokens as they arrive)
/// for try await token in session.send("What is Swift?") {
///     print(token, terminator: "")
/// }
///
/// // Await full response
/// let reply = try await session.sendAsync("Summarise the above.")
///
/// // Multi-turn — history is automatic
/// session.clearHistory()   // reset conversation, model stays loaded
/// await session.close()    // unload model, free all resources
/// ```
///
/// ## Concurrency contract
///
/// - `ChatSession` is an `actor` — `send()` calls are automatically serialized.
///   Calling `send()` while a previous stream is active queues after it completes.
/// - `cancel()` is `nonisolated` — callable from any thread/task without `await`.
///   It sets an atomic flag; the active stream drains and completes with `.cancelled`.
/// - `close()` is actor-isolated and idempotent — safe to call multiple times.
///   It waits for any active stream to finish before releasing native resources.
/// - `deinit` calls `close()` automatically — no resource leaks on deallocation.
/// - All methods throw `DeviceAiError.sessionClosed` after `close()` is called.
///
/// ## Kotlin parallel
///
/// Mirrors `ChatSession` in kotlin/llm — identical public API surface.
/// `send()` returns `AsyncThrowingStream` instead of `Flow<String>`.
/// `sendAsync()` replaces `sendBlocking()` — same semantics, Swift concurrency.
public actor ChatSession {

    private let config:  ChatConfig
    private let engine:  LlamaEngine
    private var history: [ChatTurn] = []
    private var closed = false

    init(modelPath: String, config: ChatConfig) {
        self.config = config
        self.engine = LlamaEngine()
        Task { try? await self.engine.load(modelPath: modelPath, config: config) }
    }

    // ── Read-only state ───────────────────────────────────────────────────────

    /// Completed conversation turns (user + assistant pairs).
    /// In-progress streaming turns are not included until the stream completes.
    public var turns: [ChatTurn] { history }

    /// `true` once `close()` has been called.
    public var isClosed: Bool { closed }

    // ── Messaging ─────────────────────────────────────────────────────────────

    /// Stream a response token by token.
    ///
    /// History is updated automatically:
    /// - User turn added immediately.
    /// - Assistant turn appended when the stream completes successfully.
    /// - User turn rolled back on error, allowing a clean retry.
    public func send(_ text: String, overrideConfig: ChatConfig? = nil) -> AsyncThrowingStream<String, Error> {
        AsyncThrowingStream { continuation in
            Task {
                do {
                    try self.assertOpen()
                    let userTurn = ChatTurn(role: .user, content: text)
                    self.history.append(userTurn)

                    let messages = self.buildMessages(overrideConfig: overrideConfig)
                    let genConfig = (overrideConfig ?? self.config).toGenConfig()
                    var reply = ""

                    for try await token in self.engine.generateStream(messages: messages, config: genConfig) {
                        reply += token
                        continuation.yield(token)
                    }

                    self.history.append(ChatTurn(role: .assistant, content: reply))
                    continuation.finish()
                } catch {
                    // Roll back user turn so caller can retry cleanly
                    if self.history.last?.isUser == true { self.history.removeLast() }
                    continuation.finish(throwing: error)
                }
            }
        }
    }

    /// Await the full response (non-streaming).
    /// Equivalent to collecting the entire `send()` stream.
    public func sendAsync(_ text: String, overrideConfig: ChatConfig? = nil) async throws -> String {
        var full = ""
        for try await token in send(text, overrideConfig: overrideConfig) { full += token }
        return full
    }

    // ── Lifecycle ─────────────────────────────────────────────────────────────

    /// Cancel any in-progress generation. Safe to call from any thread.
    /// The active stream completes with `.cancelled` — no error thrown to the caller.
    public nonisolated func cancel() { engine.cancel() }

    /// Clear conversation history. The model stays loaded and the session remains usable.
    public func clearHistory() { history.removeAll() }

    /// Unload the model and release all native resources.
    ///
    /// **Idempotent** — calling multiple times is safe.
    /// All subsequent method calls throw `DeviceAiError.sessionClosed`.
    public func close() async {
        guard !closed else { return }
        closed = true
        await engine.close()
        DeviceAiLogger.info("ChatSession", "Closed.")
    }

    // NOTE: async close() cannot be awaited from deinit (SE-0371).
    // cancel() is nonisolated+synchronous — stops any in-flight C++ work.
    // The engine's own deinit releases native resources.
    deinit { engine.cancel() }

    // ── Private ───────────────────────────────────────────────────────────────

    private func assertOpen() throws {
        guard !closed else { throw DeviceAiError.sessionClosed }
    }

    private func buildMessages(overrideConfig: ChatConfig?) -> [LlmMessage] {
        let cfg = overrideConfig ?? config
        var messages: [LlmMessage] = [.system(cfg.systemPrompt)]
        messages += history.map { LlmMessage(role: $0.role, content: $0.content) }
        return messages
    }
}

// ── ChatConfig → engine config mapping ───────────────────────────────────────

private extension ChatConfig {
    func toGenConfig() -> LlmGenConfig {
        LlmGenConfig(
            maxTokens:    maxTokens,
            temperature:  temperature,
            topP:         topP,
            topK:         topK,
            repeatPenalty: repeatPenalty
        )
    }
}

File: kotlin/llm/src/commonMain/kotlin/dev/deviceai/llm/ChatSession.kt
package dev.deviceai.llm

import dev.deviceai.core.InternalDeviceAiApi
import dev.deviceai.core.telemetry.InferenceEvent
import dev.deviceai.core.telemetry.TelemetryReporter
import dev.deviceai.models.currentTimeMillis
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.onCompletion
import kotlinx.coroutines.flow.onEach
import kotlin.uuid.ExperimentalUuidApi
import kotlin.uuid.Uuid

/**
 * A stateful LLM conversation session.
 *
 * Create via [dev.deviceai.core.DeviceAI.llm]:
 * ```kotlin
 * val session = DeviceAI.llm.chat("/path/to/model.gguf") {
 *     systemPrompt = "You are a helpful assistant."
 *     temperature  = 0.8f
 * }
 * ```
 *
 * ## Streaming (recommended for UI)
 * ```kotlin
 * session.send("What is Kotlin?").collect { token -> print(token) }
 * ```
 *
 * ## Blocking (scripts / tests)
 * ```kotlin
 * val reply = session.sendBlocking("What is Kotlin?")
 * ```
 *
 * ## Multi-turn — history is automatic
 * ```kotlin
 * session.send("What is Kotlin?").collect { print(it) }
 * session.send("Give me an example.").collect { print(it) }  // remembers context
 * ```
 *
 * ## Lifecycle
 * ```kotlin
 * session.cancel()        // abort in-progress generation
 * session.clearHistory()  // start a fresh conversation, keep the model loaded
 * session.close()         // unload the model and free all resources
 * ```
 */
@OptIn(InternalDeviceAiApi::class, ExperimentalUuidApi::class)
class ChatSession internal constructor(
    modelPath: String,
    private val config: ChatConfig,
) {
    /** `true` if the model loaded successfully and the session is ready for inference. */
    val isReady: Boolean = LlmCppBridge.initLlm(modelPath, config.toInitConfig())

    // Model filename without path or extension — used as telemetry model ID.
    private val modelId: String = modelPath.substringAfterLast('/').substringBeforeLast('.')

    private val _history = mutableListOf<LlmMessage>()

    /**
     * Read-only conversation history as clean [ChatTurn] pairs.
     * Only includes completed exchanges (user + assistant).
     * Excludes in-progress streaming turns.
     */
    val history: List<ChatTurn>
        get() {
            val turns = mutableListOf<ChatTurn>()
            var i = 0
            while (i + 1 < _history.size) {
                val msg = _history[i]
                val reply = _history[i + 1]
                if (msg.role == LlmRole.USER && reply.role == LlmRole.ASSISTANT) {
                    turns.add(ChatTurn(user = msg.content, assistant = reply.content))
                }
                i += 2
            }
            return turns
        }

    /**
     * Send a user message and receive a streaming response token by token.
     *
     * Conversation history is updated automatically:
     * - The user message is added immediately.
     * - The assistant reply is appended once the flow completes.
     * - If the flow errors, the user message is rolled back so the caller can retry.
     *
     * Override [config] for this request only without affecting the session default:
     * ```kotlin
     * session.send("Tell me a joke.", overrideConfig = ChatConfig().apply { temperature = 1.2f })
     * ```
     *
     * @param text           The user's message.
     * @param overrideConfig Per-request config override. Null = use session default.
     * @return [Flow] emitting token strings as they are generated.
     */
    fun send(text: String, overrideConfig: ChatConfig? = null): Flow<String> {
        _history.add(LlmMessage(LlmRole.USER, text))

        val messages = buildList {
            add(LlmMessage(LlmRole.SYSTEM, config.systemPrompt))
            addAll(_history)
        }

        val genConfig = (overrideConfig ?: config).toGenConfig()
        val reply = StringBuilder()

        val startMs = currentTimeMillis()
        val genId = Uuid.random().toString()
        var tokenCount = 0

        return LlmCppBridge.generateStream(messages, genConfig)
            .onEach { token ->
                reply.append(token)
                tokenCount++
            }
            .onCompletion { error ->
                if (error == null && reply.isNotEmpty()) {
                    _history.add(LlmMessage(LlmRole.ASSISTANT, reply.toString()))
                } else if (error != null) {
                    _history.removeLastOrNull() // roll back user message for clean retry
                }
                val totalMs = currentTimeMillis() - startMs
                TelemetryReporter.record(InferenceEvent(
                    generationId   = genId,
                    modelId        = modelId,
                    inputTokens    = estimateInputTokens(messages),
                    outputTokens   = tokenCount,
                    totalMs        = totalMs,
                    tokensPerSecond = if (totalMs > 0) tokenCount * 1000f / totalMs else 0f,
                    success        = error == null,
                    timestampMs    = currentTimeMillis(),
                ))
            }
    }

    /**
     * Send a user message and block until the full response is available.
     *
     * Prefer [send] for streaming UIs. Use this for scripts or tests.
     *
     * @param text           The user's message.
     * @param overrideConfig Per-request config override. Null = use session default.
     * @return The complete assistant response.
     */
    fun sendBlocking(text: String, overrideConfig: ChatConfig? = null): String {
        _history.add(LlmMessage(LlmRole.USER, text))

        val messages = buildList {
            add(LlmMessage(LlmRole.SYSTEM, config.systemPrompt))
            addAll(_history)
        }

        val genConfig = (overrideConfig ?: config).toGenConfig()
        val startMs = currentTimeMillis()

        return try {
            val result = LlmCppBridge.generate(messages, genConfig)
            _history.add(LlmMessage(LlmRole.ASSISTANT, result.text))
            TelemetryReporter.record(InferenceEvent(
                generationId    = Uuid.random().toString(),
                modelId         = modelId,
                inputTokens     = estimateInputTokens(messages),
                outputTokens    = result.tokenCount,
                totalMs         = result.generationTimeMs,
                tokensPerSecond = if (result.generationTimeMs > 0)
                    result.tokenCount * 1000f / result.generationTimeMs else 0f,
                success         = true,
                timestampMs     = currentTimeMillis(),
            ))
            result.text
        } catch (e: Exception) {
            _history.removeLastOrNull() // roll back user message for clean retry
            TelemetryReporter.record(InferenceEvent(
                generationId    = Uuid.random().toString(),
                modelId         = modelId,
                inputTokens     = estimateInputTokens(messages),
                outputTokens    = 0,
                totalMs         = currentTimeMillis() - startMs,
                tokensPerSecond = 0f,
                success         = false,
                timestampMs     = currentTimeMillis(),
            ))
            throw e
        }
    }

    /** Abort any in-progress [send] or [sendBlocking] call. */
    fun cancel() = LlmCppBridge.cancelGeneration()

    /** Clear conversation history. The model stays loaded and the session remains usable. */
    fun clearHistory() = _history.clear()

    /** Unload the model and release all engine resources. Do not use the session after this. */
    fun close() = LlmCppBridge.shutdown()

    // ── Helpers ───────────────────────────────────────────────────────────────

    /** Rough token count estimate — 1 token ≈ 4 chars for English text. */
    private fun estimateInputTokens(messages: List<LlmMessage>): Int =
        messages.sumOf { it.content.length } / 4
}
